source/Mlos.Python/mlos/Optimizers/ExperimentDesigner/UtilityFunctions/SeriesDifferenceUtilityFunction.py

```diff
@@ -0,0 +1,98 @@
+            CategoricalDimension(name="utility_function_name", values=["lower_confidence_bound_on_improvement", "upper_confidence_bound_on_improvement"]),
+            ContinuousDimension(name="alpha", min=0.01, max=0.5)
+        ]
+    ),
+    default=Point(
+        utility_function_name="upper_confidence_bound_on_improvement",
+        alpha=0.01
+    )
+)
+
+
+class SeriesDifferenceUtilityFunction(UtilityFunction):
+    def __init__(self, function_config: Point, surrogate_model: MultiObjectiveRegressionModel, objective: SeriesObjective, logger=None):
+        if logger is None:
+            logger = create_logger(self.__class__.__name__)
+        self.logger = logger
+
+        self.config = function_config
+        self.series_objective = objective
+        self._sign = 1 if not self.series_objective.minimize else -1
+        self.surrogate_model: MultiObjectiveRegressionModel = surrogate_model
+
+    @trace()
+    def __call__(self, feature_values_pandas_frame):
+        # NOTE: THIS LOGIC IS DUPLICATED INTO SeriesPredictedValueUtilityFunction and SeriesDifferenceUtilityFunction.
+        # Later they will diverge as SeriesDifferenceUtilityFunction should require additional logic surrounding probability
+        # of improvement. Perhaps this part should be refactored to be shared though.
+        #
+        self.logger.debug(f"Computing utility values for {len(feature_values_pandas_frame.index)} points.")
+
+        predicted_value_col = Prediction.LegalColumnNames.PREDICTED_VALUE.value
+        predicted_value_var_col = Prediction.LegalColumnNames.PREDICTED_VALUE_VARIANCE.value
+        dof_col = Prediction.LegalColumnNames.PREDICTED_VALUE_DEGREES_OF_FREEDOM.value
+
+        del feature_values_pandas_frame[f"context_space.{self.series_objective.series_modulation_dimension.name}"]
+        series_vals_df = pd.DataFrame({
+            f"context_space.{self.series_objective.series_modulation_dimension.name}": self.series_objective.series_modulation_dimension.linspace()
+        })
+        feature_values_pandas_frame_merged = feature_values_pandas_frame.merge(series_vals_df, how="cross")
+        multi_objective_predictions = self.surrogate_model.predict(features_df=feature_values_pandas_frame_merged)
+
+        predictions = multi_objective_predictions[0]
+        predictions_df = predictions.get_dataframe()
+
+        # TODO ZACK: Vectorize this. I'm sure pandas/numpy can do this faster. @Adam, any suggestions
+        resulting_series_arry = []
+        current_arry = []
+        for index, prediction in predictions_df.iterrows():
+            current_arry.append(prediction[predicted_value_col])
+            if ((index + 1) % len(self.series_objective.series_modulation_dimension)) == 0:
+                resulting_series_arry.append(np.array(current_arry))
+                current_arry=[]
+
+        # Didn't do bootstrapping yet because I think math would be computationally faster

```

Comment:
The for-loop above is very brittle. Use the 'index column' to group the predictions by the features that created them. And remember to restore the index so that the caller can properly match the utility function value, to the features.

Reply:
Good idea.